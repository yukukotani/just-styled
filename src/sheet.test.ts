import { describe, expect, test } from "vitest";
import {
	generateSheets,
	generateThemeSheets,
	resolvePropertyValue,
	transformTokenToCssVariable,
} from "./sheet";

describe("generateSheets", () => {
	test("generates simple single property", () => {
		const sheets = generateSheets({
			color: "red",
		});

		expect(sheets).toHaveLength(1);
		expect(sheets).toMatchSnapshot();
	});

	test("generates camelCase property as kebab-case key", () => {
		const sheets = generateSheets({
			backgroundColor: "black",
		});

		expect(sheets).toHaveLength(1);
		expect(sheets).toMatchSnapshot();
	});

	test("generates multiple properties", () => {
		const sheets = generateSheets({
			border: "1px solid red",
			borderRadius: "4px",
		});

		expect(sheets).toHaveLength(2);
		expect(sheets).toMatchSnapshot();
	});

	test("generates the same class for the same style", () => {
		const sheets1 = generateSheets({
			color: "red",
		});
		const sheets2 = generateSheets({
			color: "red",
		});

		expect(sheets1).toHaveLength(1);
		expect(sheets2).toHaveLength(1);
		expect(sheets1[0].className === sheets2[0].className).toBe(true);
	});

	test("generates different classes for different values", () => {
		const sheets1 = generateSheets({
			color: "red",
		});
		const sheets2 = generateSheets({
			color: "white",
		});

		expect(sheets1).toHaveLength(1);
		expect(sheets2).toHaveLength(1);
		expect(sheets1[0].className === sheets2[0].className).toBe(false);
	});

	test("generates different classes for different properties", () => {
		const sheets1 = generateSheets({
			color: "red",
		});
		const sheets2 = generateSheets({
			backgroundColor: "red",
		});

		expect(sheets1).toHaveLength(1);
		expect(sheets2).toHaveLength(1);
		expect(sheets1[0].className === sheets2[0].className).toBe(false);
	});

	test("generates custom property reference", () => {
		const sheets = generateSheets({
			color: "$colors.red",
		});

		expect(sheets).toHaveLength(1);
		expect(sheets).toMatchSnapshot();
	});

	test("generates custom property reference with kebab-case transformation", () => {
		const sheets = generateSheets({
			color: "$colors.lightGray",
		});

		expect(sheets).toHaveLength(1);
		expect(sheets).toMatchSnapshot();
	});
});

describe("generateThemeSheets", () => {
	test("generates single token", () => {
		const sheet = generateThemeSheets({
			tokens: {
				colors: {
					lightGray: "#666666",
				},
			},
		});

		expect(sheet).toMatchSnapshot();
	});

	test("generates multiple tokens", () => {
		const sheet = generateThemeSheets({
			tokens: {
				colors: {
					"gray.200": "#eeeeee",
					"gray.600": "#666666",
				},
				spaces: {
					sm: "4px",
				},
				sizes: {
					lg: "120px",
				},
				fontSizes: {
					md: "16px",
				},
				radii: {
					xs: "2px",
				},
			},
		});

		expect(sheet).toMatchSnapshot();
	});

	test("generates a token which refer the other token", () => {
		const sheet = generateThemeSheets({
			tokens: {
				colors: {
					gray: "#666666",
					primary: "$colors.gray",
				},
			},
		});

		expect(sheet).toMatchSnapshot();
	});
});

describe("transformTokenToCssVariable", () => {
	test("keeps lowercase value", () => {
		const res = transformTokenToCssVariable("colors.myred");
		expect(res).toBe("--j-colors-myred");
	});

	test("keeps kebab-case value", () => {
		const res = transformTokenToCssVariable("colors.my-red");
		expect(res).toBe("--j-colors-my-red");
	});

	test("transforms dot-separated value", () => {
		const res = transformTokenToCssVariable("colors.red.500");
		expect(res).toBe("--j-colors-red-500");
	});

	test("transforms lowerCamelCase value", () => {
		const res = transformTokenToCssVariable("colors.lightGray");
		expect(res).toBe("--j-colors-light-gray");
	});

	test("transforms value with category parameter", () => {
		const res = transformTokenToCssVariable("red.500", "colors");
		expect(res).toBe("--j-colors-red-500");
	});

	test("transforms complex value", () => {
		const res = transformTokenToCssVariable("colors.lightGray.500");
		expect(res).toBe("--j-colors-light-gray-500");
	});
});

describe("resolvePropertyValue", () => {
	test("keeps string value", () => {
		const res = resolvePropertyValue("red");
		expect(res).toBe("red");
	});

	test("resolves token", () => {
		const res = resolvePropertyValue("$colors.red.500");
		expect(res).toBe("var(--j-colors-red-500)");
	});
});
