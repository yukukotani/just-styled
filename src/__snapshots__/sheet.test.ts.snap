// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateSheets > generates a combinator 1`] = `
[
  {
    "className": "j1dqz",
    "css": ".j1dqz { color: red; }",
  },
  {
    "className": "jiy1a",
    "css": ".jiy1a> span { color: purple; }",
  },
]
`;

exports[`generateSheets > generates a media query 1`] = `
[
  {
    "className": "j15vw",
    "css": "@media screen and (width: 600px){ .j15vw { color: plum; } }",
  },
]
`;

exports[`generateSheets > generates a nested pseudo selector 1`] = `
[
  {
    "className": "j1dqz",
    "css": ".j1dqz { color: red; }",
  },
  {
    "className": "jm8xk",
    "css": ".jm8xk:hover { color: blue; }",
  },
  {
    "className": "je9tx",
    "css": ".je9tx:hover:active { background-color: pink; }",
  },
]
`;

exports[`generateSheets > generates a pseudo selector 1`] = `
[
  {
    "className": "j1dqz",
    "css": ".j1dqz { color: red; }",
  },
  {
    "className": "jm8xk",
    "css": ".jm8xk:hover { color: blue; }",
  },
]
`;

exports[`generateSheets > generates an attribute selector 1`] = `
[
  {
    "className": "j1dqz",
    "css": ".j1dqz { color: red; }",
  },
  {
    "className": "j12qb",
    "css": ".j12qb[data-foo="bar"] { color: yellow; }",
  },
]
`;

exports[`generateSheets > generates an selector in a media query 1`] = `
[
  {
    "className": "ji344",
    "css": "@media screen and (width: 600px){ .ji344:hover { color: teal; } }",
  },
]
`;

exports[`generateSheets > generates an selector in a media query 2`] = `
[
  {
    "className": "j1msw",
    "css": "@media screen and (width: 600px){ .j1msw:hover { color: tomato; } }",
  },
]
`;

exports[`generateSheets > generates camelCase property as kebab-case key 1`] = `
[
  {
    "className": "jnqp7",
    "css": ".jnqp7 { background-color: black; }",
  },
]
`;

exports[`generateSheets > generates custom property reference 1`] = `
[
  {
    "className": "j1ovh",
    "css": ".j1ovh { color: var(--j-colors-red); }",
  },
]
`;

exports[`generateSheets > generates custom property reference with kebab-case transformation 1`] = `
[
  {
    "className": "j10nw",
    "css": ".j10nw { color: var(--j-colors-light-gray); }",
  },
]
`;

exports[`generateSheets > generates multiple properties 1`] = `
[
  {
    "className": "j1pim",
    "css": ".j1pim { border: 1px solid red; }",
  },
  {
    "className": "jr2ve",
    "css": ".jr2ve { border-radius: 4px; }",
  },
]
`;

exports[`generateSheets > generates simple single property 1`] = `
[
  {
    "className": "j1dqz",
    "css": ".j1dqz { color: red; }",
  },
]
`;

exports[`generateThemeSheets > generates a token which refer the other token 1`] = `":root { --j-colors-gray: #666666; --j-colors-primary: var(--j-colors-gray); }"`;

exports[`generateThemeSheets > generates multiple tokens 1`] = `":root { --j-colors-gray-200: #eeeeee; --j-colors-gray-600: #666666; --j-spaces-sm: 4px; --j-sizes-lg: 120px; --j-font-sizes-md: 16px; --j-radii-xs: 2px; }"`;

exports[`generateThemeSheets > generates single token 1`] = `":root { --j-colors-light-gray: #666666; }"`;
